src/runtime-rs/crates/hypervisor/src/device/topology.rs

```diff
@@ -0,0 +1,316 @@
+   -----------   ------------------   -------------------   --------------
+
+Plug only PCI Express devices into PCI Express Ports.
+
+   pcie.0 bus
+   ----------------------------------------------------------------------------------
+        |                 |                                    |
+   -------------    -------------                        -------------
+   | Root Port |    | Root Port |                        | Root Port |
+   ------------     -------------                        -------------
+         |                            -------------------------|------------------------
+    ------------                      |                 -----------------              |
+    | PCIe Dev |                      |    PCI Express  | Upstream Port |              |
+    ------------                      |      Switch     -----------------              |
+                                      |                  |            |                |
+                                      |    -------------------    -------------------  |
+                                      |    | Downstream Port |    | Downstream Port |  |
+                                      |    -------------------    -------------------  |
+                                      -------------|-----------------------|------------
+                                             ------------
+                                             | PCIe Dev |
+                                             ------------
+*/
+
+use std::{
+    collections::HashMap,
+    sync::{Arc, RwLock},
+};
+
+use crate::{HYPERVIOR_CLH, HYPERVISOR_DRAGONBALL};
+
+use super::pci_path::get_pcipath_from_bdf;
+
+// add device to topology.
+// let do_add_pcie_device =
+//     |dev: Box<dyn PCIeDevice>| -> Option<u32> { self.pcie_topology.insert_device(dev, None) };
+#[macro_export]
+macro_rules! do_add_device {
+    ($pcie_topology:expr, $opt:expr) => {
+        |dev: Box<dyn PCIeDevice>| -> Option<u32> {
+            // When VMM can return the PCI address,we just get it and insert the device into PCI topology
+            $pcie_topology.insert_device(dev, $opt)
+        }
+    };
+}
+
+// remove device from topology.
+// let do_add_pcie_device =
+//     |device_id: String| -> Option<u32> { self.pcie_topology.remove_device(device_id) };
+#[macro_export]
+macro_rules! do_remove_device {
+    ($pcie_topology:expr) => {
+        |device_id: String| -> Option<u32> { $pcie_topology.remove_device(device_id.as_str()) }
+    };
+}
+
+// PCIeDevice trait
+pub trait PCIeDevice: Send + Sync {
+    fn device_type(&self) -> PCIeDevType;
+    fn device_id(&self) -> &str;
+}
+
+// 00/FF not included
+const PCI_BRIDGE_MAX_CAPACITY: u32 = 30;
+
+pub const DRIVER_MMIO_BLK_TYPE: &str = "mmioblk";
+pub const DRIVER_VFIO_PCI_TYPE: &str = "vfio-pci";
+const DEFAULT_PCIE_ROOT_BUS: &str = "pcie.0";
+const DEFAULT_PCIE_ROOT_BUS_ADDRESS: &str = "0000:00";
+
+#[derive(Clone, Debug, Default, PartialEq)]
+pub enum BusMode {
```

Comment:
since this bus mod only related with vfio, so is it better to keep it in vfio.rs?

Reply:
Thx @lifupan 
I'll need to take a second look at this, because bus mode is also a factor now that we're considering PCI topology.